dll/win32/shlwapi/propbag.cpp

```diff
@@ -1188,3 +1188,359 @@ HRESULT SHGetDesktopUpgradePropertyBag(REFIID riid, void **ppvObj)
+        return SHGetShellKey((SHKEY_Key_Shell | SHKEY_Root_HKCU), NULL, TRUE);
+
+    return SHGetShellKey(SHKEY_Key_ShellNoRoam | SHKEY_Root_HKCU, NULL, TRUE);
+}
+
+HRESULT CViewStatePropertyBag::_GetRegKey(
+    LPCITEMIDLIST pidl,
+    LPCWSTR pszBagName,
+    DWORD dwFlags,
+    char unknown,
+    HKEY hKey,
+    LPWSTR pszDest,
+    INT cchDest)
+{
+    FIXME("Stub\n");
+    return E_NOTIMPL;
+}
+
+HRESULT
+CViewStatePropertyBag::_CreateBag(
+    LPITEMIDLIST pidl,
+    LPCWSTR pszPath,
+    DWORD dwVspbFlags,
+    DWORD dwMode,
+    REFIID riid,
+    void **ppvObj)
+{
+    FIXME("Stub\n");
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP
+CViewStatePropertyBag::Read(
+    _In_z_ LPCWSTR pszPropName,
+    _Inout_ VARIANT *pvari,
+    _Inout_opt_ IErrorLog *pErrorLog)
+{
+    FIXME("Stub\n");
+    return E_NOTIMPL;
+}
+
+static BOOL SHIsRemovableDrive(LPCITEMIDLIST pidl)
+{
+    STRRET strret;
+    CComPtr<IShellFolder> psf;
+    WCHAR szBuff[MAX_PATH];
+    LPCITEMIDLIST ppidlLast;
+    INT iDrive, nType;
+    HRESULT hr;
+
+    hr = SHBindToParent(pidl, IID_IShellFolder, (void **)&psf, &ppidlLast);
+    if (FAILED(hr))
+        return FALSE;
+
+    hr = psf->GetDisplayNameOf(ppidlLast, SHGDN_FORPARSING, &strret);
+    if (FAILED(hr))
+        return FALSE;
+
+    hr = StrRetToBufW(&strret, ppidlLast, szBuff, _countof(szBuff));
+    if (FAILED(hr))
+        return FALSE;
+
+    iDrive = PathGetDriveNumberW(szBuff);
+    if (iDrive < 0)
+        return FALSE;
+
+    nType = RealDriveType(iDrive, FALSE);
+    return (nType == DRIVE_REMOVABLE || nType == DRIVE_CDROM);
+}
+
+/**************************************************************************
+ *  SHGetViewStatePropertyBag (SHLWAPI.515)
+ *
+ * Retrieves a property bag in which the view state information of a folder
+ * can be stored.
+ *
+ * @param pidl      PIDL of the folder requested
+ * @param bag_name  Name of the property bag requested
+ * @param flags     Optional flags
+ * @param riid      IID of requested property bag interface
+ * @param ppv       Address to receive pointer to the new interface
+ * @return          An HRESULT value. S_OK on success, non-zero on failure.
+ * @see             https://learn.microsoft.com/en-us/windows/win32/api/shlwapi/nf-shlwapi-shgetviewstatepropertybag
+ */
+EXTERN_C HRESULT WINAPI
+SHGetViewStatePropertyBag(
+    _In_opt_ PCIDLIST_ABSOLUTE pidl,
+    _In_opt_ LPCWSTR bag_name,
+    _In_ DWORD flags,
+    _In_ REFIID riid,
+    _Outptr_ void **ppv)
+{
+    HRESULT hr;
+
+    *ppv = NULL;
+
+    ::EnterCriticalSection(&g_csBagCacheLock);
+
+    if (g_pCachedBag && g_pCachedBag->IsSameBag(pidl, bag_name, flags))
+    {
+        hr = g_pCachedBag->QueryInterface(riid, ppv);
+        ::LeaveCriticalSection(&g_csBagCacheLock);
+        return hr;
+    }
+
+    if (SHIsRemovableDrive(pidl))
+    {
+        ::LeaveCriticalSection(&g_csBagCacheLock);
+        return E_FAIL;
+    }
+
+    {
```

Comment:
Why those extra parentheses? Is it for the pBag destruction to happen before the critical section unlock?

Reply:
Removed.